### DISCUSSION TITLE
GADTs

### DISCUSSION THREAD

**Discussion by sirinath** [dotnet/csharplang#5527] [Language Ideas]
Other languages like Scala, Ceylon, F# having some form of ADT, is it possible to consider adding GADT to C# so we can write something like:

```csharp
class C<X, Y, Z> = A<X<Z<P _, P _>> | B<Y<_>> 
```

Maybe `type` can be made a keyword in which case this would be:

```csharp
type C<X, Y, Z> = A<X<Z<_ P, _ P>> | B<Y<_>> 
```

where wide card parameter to `Z` is contained to be the same type and the wild car parameters to `Y` has no such constraints.

**Comment by YairHalberstadt** [comment]
It's not clear what you're referring to by GADT?

**Reply by HaloFour** [reply]
I believe it's referring to this: https://en.wikipedia.org/wiki/Generalized_algebraic_data_type

**Comment by orthoxerox** [comment]
What are some use cases that are enabled by GADTs that cannot be implemented sufficiently elegantly in C#11 (or C#11 + ADTs)?

**Reply by daiplusplus** [reply]
Do you have a link to the current state of ADTs for C# 11? I haven't seen anything concrete in a while and what proposals I did see weren't true ADTs nor GADTs).

I understand [that GADTs are a superset of ADTs: just like ADTs but with the ability to be self-referential](https://cstheory.stackexchange.com/questions/10594/whats-the-difference-between-adts-gadts-and-inductive-types) - given the limited expressiveness of non-recursive ADTs my post below treats "ADT" and "GADT" as synonyms.

<sub>Also, I apologize in advance if I'm misusing category-theory terms (like _monomorphism_). Please correct me wherever I use incorrect terminology.</sub>

----------------

Assuming that the C# 11 story doesn't look too different from C# 10.0 then I'm really hoping for true ADTs to significantly boost my QoL and productivity because it would allow me to be expressive without being verbose, and enable some things that are just _impractical_ or extremely tedious and error-prone in C# today.

-----------------------

## _True Sum types_

* aka union-types, though not necessarily the same as discriminated-unions (aka tagged-unions).
  * (hat-tip to @HaloFour for correcting me on this distinction last week).
* In TypeScript an example would be `number | string` or `Customer | CustomerLookupError`.
* Currently the best we have is @mcintyre321 's excellent `OneOf<...>` library, though it has limitations imposed on it by the rest of C#/.NET...
    * ...such as not being able to "inherit" or "mixin" common members of all types (so a `OneOf<Customer,Contact>` won't have a `String FirstName { get; }` property - we'd have to either add extension-methods (which aren't properties!) or use `Match`/`Switch` for every type, which is tedious.
    * There's no concept of type-equivalence: `OneOf<Customer,Order>` is considered distinct from `OneOf<Order,Customer>`. This can be worked-around by adding `implicit` conversions between `OneOf<...>` types with all the different permutations of type-arguments, but once you go beyond 4 type-parameters it becomes unwieldly.
    * Type parameters cannot be constrained relative to the others, so `OneOf<Int64,Int64,Int64,String>` is valid, even though it should always be reduced to `OneOf<Int64,String>` (again, you can have `implicit` conversions to safely perform reduction, but it isn't elegant).
    *  Due to how value-types are packed in .NET a `OneOf<...>` (without using `unsafe` code) with _N-many_ type-parameters will need N-many fields even though only 1 will ever be used, so a `OneOf<Int64,Decimal,UInt64,SomeStruct,String>` will use at least 48 bytes which has performance implications.
* Type-unions are also an effective (if not superior) alternative to inheritance and `virtual` methods for runtime polymorphism, and let you side-step issues with `sealed` types and members.

## _True Product Types_

* aka Aggregates, aka Tuples (in a mathematical sense)
* Traditionally, and strictly-speaking, every `class` and `struct` in C#/.NET with 2 or more fields is a _product type_.
  * However, there are far too many restrictions, limitations and implications of .NET `class` and `struct` types for them to be considered _true_ ADTs:
    * C#/.NET's type-system has no concept of type-equivalence for separate `class`/`struct` types nor has any kind of built-in support for _type isomorphisms_  (while you can hack-together [a poor-man's version of `reinterpret_cast` over value-types using IL Emit](https://github.com/dotnet/runtime/discussions/58535), the compiler will not warn you about unsafe or impossible casts (at least, not without a custom analyzer rule).
    * Additionally, implementing a `struct` inevitably forces you into overriding `GetHashCode`, `Equals`, adding the equality and comparison operators, and more besides. This is very, very tedious as C# doesn't have a spaceship-operator `<=>` nor does C# implement equality and comparison operators for you if you already implement `IComparable`. This is silly and needs fixing.
    * Yes, C#'s `record` types do go a long way to address the problems of `class`/`struct` types, they don't go far enough: So all `record` types really do is save you from having to repeatedly type-out an immutable type's members 3 times over (1: constructor parameter, 2: public property definition, and 3: assigning the property in the constructor). They still aren't composable or easily extendable.
  * While `ValueTuple<...>` superficially resembles a product-type, it isn't a _safe_ product-type because every `ValueTuple`'s members' names are erased at compile-time and don't participate in nominative typing. While names are preserved in attributes, they don't factor into the compiler's type-safety decision-making. So if you have a method `GetUserInfo` with a value-tuple return-type that's `( Int32 userId, String bio )` and another method `WipeLocalHardDrive` that accepts a value-tuple parameter `( Int32 timeToAbortMilliseconds, String bitlockerPassword )` then the compiler won't stop you from passing that user-info tuple as a parameter for `WipeLocalHardDrive` - and _yes_, issues like that _have caused, and continue to cause_ issues in production-code (though I'll admit the real cause is how Entity Framework doesn't support custom struct-types to represent `int` `PRIMARY KEY` values, that way `Int32 userId` would actually be `UserKey userId` and it couldn't be passed into `WipeLocalHardDrive` in our contrived example.
  * Another limitation today is that product-types cannot be variant over their types - though a workaround exists by hand-implementing every possible implicit conversion. While `interface` types can represent product-types and generic interfaces _can_ be covariant and contravariant over their type-parameters this does not extend to non-generic interfaces; additionally interface type parameter variance cannot be used with `out` params which is an extremely unfortunate situation.

## _True Intersection Types_

* A TypeScript example could be `Customer & ExtraCustomerData`.
* C#/.NET (technically) supports two distinct (and **very** limited) kinds of intersection-types:
   1. Class inheritance (insofar as `class Derived : Base` can be thought-of as `Derived & Base`).
   1. Interface intersection, but only in generic type-parameter constraints (e.g. `Foo<T>( T value ) where T : ISomeInterface, IAnotherInterface`).
       * I struggle to find a legitimate use-case for `interface` intersection constraints for parameter types because ultimately the _concrete type argument_ will be a non-interface type `TActual` that implements all interfaces, and if you have that then you might-as-well just make the method non-generic and use `TActual` as the parameter-type - or make your own single interface type and ditch the intersection. And I haven't found any use for interface intersection types as return-types (this is my personal experience, I'm sure other people have found uses for it though).
* _True product types_ would finally eliminate one of my biggest complaints about the C#/.NET type-system: the inability to incrementally _add_ state to an existing object of another existing type, and especially adding state to objects of `sealed` types with _dozens or more_ data-members (Entity Framework strikes again...).
  * Consider a `class Order` (representing a customer purchase order on an online ecommerce site). Supposing that you want to "attach" extra data to an `Order` object, currently you have a few options, none of which are particularly compelling:
    1. Subclass `class Order` and add the new state (e.g. `class Order2 : Order`). But this requires you to either add a "copy constructor",  (because you have an existing `Order` object you want to extend) which is _very tedious_ as you need to copy-over every data-member, or update wherever-it-is in your program to change `new Order(...)` to `new Order2` assuming that even makes any sense. Though at least this way all existing consumers of `Order` can be safely given an `Order2` instance at runtime without breaking anything. Obviously this isn't an option if `class Order` is `sealed` or doesn't expose an appropriate constructor,
    2. Wrap the `Order` in a `ValueTuple` with extra data. The downside to this approach is that you now cannot pass `ValueTuple<Order,String,Int32,Etc>` to a consumer expecting `Order`, and assuming you do go through the tedium to update all those consumers to accept a `ValueTuple<Order,String,Int32,Etc>` instead you'll now find yourself with the maintenance burden of having to ensure all those `ValueTuple` signatures exactly match. Ugh. Also this approach has all the other same problems that "`ValueTuple`-as-a-product-type" has, as described above.
    3. Use the `ConditionalWeakTable`-technique to store the additional data in a dictionary that uses `WeakReference` to attach itself to the original `Order` object. This approach feels like a hack (non-local effects are bad!) and the lack of determinism in `WeakReference` makes it harder to reason about object lifetime.
    4. Define a custom `struct` wrapper which re-implements all of the properties and methods of `class Order` as pass-through/forwarded members and has the same interfaces and supports implicit conversion to `Order`. This is currently probably the most elegant approach, but also the most tedious: all additional data-members or state cannot be stored directly in the `struct` (mutable structs don't work), instead they must be stored in another distinct `class` that the `struct` wraps along with the original `class Order`. Though this approach also suffers from the same problem as `ValueTuple` in that when you pass it to a consumer of `Order` the extra data _is lost_.

## More _Real ADT_ benefits

* Sum and Intersection types offer a real alternative to `virtual` methods for runtime polymorphic behaviour, especially between types with completely incompatible interfaces.
* ADTs really shine when the language has _type operators_ or _utility types_ that allow you to define new types in reference to _aspects_ of other types, including _function types_ (not just `delegate` types).
  * TypeScript has `ReturnType<FunctionType>`, `Readonly<T>`, `Parameters<FunctionType>` (and many others), but these 3 offer _huge benefits in expressiveness_ to the programmer. I wish I had this functionality in C# all the time because it would eliminate so much error-prone tedium.
    * For example, consider this C# 10 record class: `public record class Customer( Int32 CustomerId, String FirstName, String LastName, DateTimeOffset Created, Address StreetAddress );` (and then add another 20 or so members/parameters because every CRM I've ever seen is full of them).
      * Supposing that you need to write a factory method somewhere which has the same parameters as `record class Customer` and passes those parameters over to `Customer`'s constructor after doing some custom logic. Currently you need to re-type out all of those parameters and ensure they're kept in-sync with those in `record class Customer`. When you have 20+ members, that's a lot of tedium and puts the program at risk of human-error. So you end up with `static Customer CreateCustomer( Int32 customerId, String firstName, String lastName, DateTimeOffset created, Address streetAddress, etc ) { /* do stuff */ return new Customer( CustomerId: customerId, FirstName: firstName, etc, argh, my carpel-tunnel!  )`
       * But if we could simply do `static Customer CreateCustomer( Parameters<Customer.Customer> p ) { /* do stuff */ return new Customer( p ); }` and the C# compiler would automatically expand `p` just like JavaScript's `...` spread-operator.
       * But wait, _there's more!_: you can also just add more parameters! So you can do  `static Customer CreateCustomer( Parameters<Customer.Customer> p, String extraParam1, Int32 extraParam2 ) { /* do stuff */ return new Customer( p ); }`. **This** is the kind of powerful expressiveness that I want.
* The second major benefit of _true_ intersection types (when used in conjunction with predicate functions, and with a `never` _bottom type_) is that they can be used to succinctly define and represent _refinement types_.
  * Currently, implementing refinement-types in C# requires each-and-every distinct refinement-constraint to be expressed as a distinct `struct` type which automatically brings along with it all the downsides of writing and maintaining `struct` types (I suppose you could use a `class` type too).
  * But when you have intersection-types at your disposal then you can express refinement types as the intersection of the source data-type (e.g. `string` or some `class` type) and some type-level invariant expressed in type-form.
    * Note that there are different _kinds_ of refinement-types (for example, a refinement that asserts a `string` contains a valid SMTP e-mail address; a different kind of refinement-type would assert that members of _some other type_ are `!= null`. This is very useful when working with Entity Framework's entity classes and you want to statically assert that `Customer`'s member collection `List<Order>` is fully-loaded and not `null` or empty (indeed, faulty assumptions about Entity Framework entity object member references and collections is my current top source of bugs. I currently use T4 to generate hundreds of `struct` refinement-types to assert that members are non-null, which isn't the best situation, but better than nothing).
       * For example, the `string-is-valid-email` refinement type would be `string & { isValidEmailTag: never }` (the `never` property prevents implicit conversion _from_ any `string`) while the type-guard (predicate function) performs the actual regex validation and asserts that a given `string` type-matches the refinement-type (so there isn't actually any `isValidEmail` member anywhere ).
       * For the `Customer-orders-are-loaded` the refinement type could be `Customer & { orders: List<Order> & NonEmptyCollection<Order> }` (where `NonEmptyCollection<T>` is `{ count: NonZero }` and `NonZero` is `int & { isNonZero: never }` (with a predicate/guard function to verify that `int value > 0`.

---------------------

## On implementation...

* I believe a quick-and-dirty first-draft for demo purposes could be implemented using only Roslyn code-generation and maybe some Roslyn analyzers too...
   * ...though it'd be gnarly-looking without the ability to extend C#'s syntax (such as defining ADTs inside a `const string` inside an assembly attribute, e.g. `[assembly: DeclareUnion( "String | Int32" )]`.
   * It also shouldn't require making _substantial_ changes to the .NET CLR and runtime. Code to support ADT syntax would be confined to Roslyn, and ADT syntax itself would be be compiled-down to a load of invisible value-types (this is exactly how "Extension _Everything_" was proposed), (heck it might even work with the .NET Framework 4 CLR). ADTs would be "exported" in `[assembly: ]` attributes that contain enough info to allow `csc` to generate new ADT monomorphs itself, even if the original program used C# native syntax to define ADTs.

* First-draft implementation thoughts:
  * Sum Types:
    * In short: using `OneOf<...>` as the basis of the implementation: instead of using generic `struct OneOf<T0,T1,T2,...>` code-gen would fully define a monomorphized (i.e. non-generic) struct which would allow it to contain properties and methods that forward-on to its contained values _and_ implement any shared interfaces, and so on. And code-analysis would allow automatic on-demand generation of implicit conversion operators to handle type-equivalence without needing us to generate all permutations of type-arguments ahead-of-time..
  * Product Types:
    * Again, Roslyn code-gen can be used to generate new `class` and `struct` types from some _aggregate definition source_, such as another `class`/`struct`/`interface` type, or from method parameters and return-types, or anything else.
  * Intersection Types:
    * This is the difficult one...
    * Intersection types _without state-loss_ could be initially implemented using `ConditionalWeakTable` behind-the-scenes, though I'd like to see the CLR modified to detect types used in intersections and automatically _reserve_ a single reference-type field at the end of the source-type which would be set to a reference to some ADT metadata structure.
    * Refinement types would really benefit from `csc`'s C# flow analysis support for type-guard functions without the need to introduce a new refined-value (an instance of a refinement-type) into the local scope, instead `csc` would add the asserted refinement as just another tag to its flow-analysis of the affected local/param. As a stopgap type-guard functions could be implemented as `bool TryRefine( [NotNullWhen(true)] out RefinmentType refined )` (assuming `NotNullWhen` can be used to prevent any use of the `out` parameter if the function returns `false`) - this is how I work with my refinement-types right now, actually.
      * Another thought I had was to take advantage of .NET's GC's ability to move and relocate objects in-memory to _forcibly subclass_ already-constructed objects, and maybe have fun with their vtables too.
        * For example, given an object instance `obj` of `class Base` and the class definition `class Derived : Base`, we can _promote_ `obj` to `Derived` by relocating `obj` to a newly allocated memory block that's big enough for `Derived` and running `Derived`'s constructor (but not `Base`'s constructor). Provided the types are immutable - or otherwise are such that the subtype's constructor doesn't invalidate the base type's constructor's guarantees - then this should _just work_... right?

* Other misc thoughts:
  * Another nice-to-have would be to finally allow implicit conversion from a `struct` to an `interface` type without requiring the struct to implement the interface (and thus avoid boxing) and allow `ref struct` types to be used too.
    * Though what would be even better is if the .NET runtime converted every function accepting interface-type parameters to a generic function where the interface is a constraint, so `public void Foo( ISomething foo )` becomes `public void Foo<T>( T foo ) where T : ISomething` - this would happen behind-the-scenes.

-----

Wow, dang - I spent 3 hours writing and researching that...

**Reply by Xyncgas** [reply]
Just beat it

**Reply by 333fred** [reply]
@Xyncgas this is your second warning for CoC violating-comments in the last couple of days. Telling someone to leave this community is not acceptable behavior, and further infractions will result in harsher penalties. Please make sure that you are following our [Code of Conduct](https://dotnetfoundation.org/about/code-of-conduct) on all dotnet repositories.

**Comment by radrow** [comment]
How do you imagine syntax for this? Could you provide us with some examples? I think it could be quite informative to see a C# version of the Wikipedia example:
```haskell
data Expr a where
    EBool  :: Bool     -> Expr Bool
    EInt   :: Int      -> Expr Int
    EEqual :: Expr Int -> Expr Int  -> Expr Bool

eval :: Expr a -> a
eval e = case e of
    EBool a    -> a
    EInt a     -> a
    EEqual a b -> (eval a) == (eval b)
```

**Comment by alberdingk-thijm** [comment]
This [paper](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/gadtoop.pdf) by Kennedy (@andrewjkennedy) and Russo (@crusso) from 2016 proposes GADTs for C# version 2.0 using an object-oriented approach. Sadly, while they expressed the intent to add this to the language, it never ended up happening, and neither of them are still at MSR from what I can gather.

They outline evaluation for typed expressions in Figure 2, but note that the cooler GADT features they might want to implement require extensions to C#'s type system. Here's an example demonstrating the specific `where T = E` type constraint syntax they proposed (taken from Figure 5 of the paper):

```csharp
public abstract class Exp<T> {
  // ...
  public virtual bool TupleEq<C,D>(Tuple<C,D> that) where T=Pair<C,D>
  {
    return false;
  }
}

public class Tuple<A,B> : Exp<Pair<A,B>> {
  // ...
  public override bool Eq(Exp<Pair<A,B>> that)
  { 
    return that.TupleEq<A,B>(this);
  }

  public override bool TupleEq<C,D>(Tuple<C,D> that)
  {
    return that.e1.Eq(e1) && that.e2.Eq(e2);
  }
}
```

**Reply by sirinath** [reply]
Many thanks for sharing.

The above paper mentions at the end they have made other proposals. Is it possible to share then or give a pointer to those proposals are.

**Reply by andrewjkennedy** [reply]
I don't believe any further work was done in the product.

As for further research work, you may be interested in a follow-up paper that generalized from equality to subtype constraints, as hinted at in the final paragraph of the GADT paper. See https://www.microsoft.com/en-us/research/publication/variance-and-generalized-constraints-for-c-generics/

Of course, variance was implemented in both the CLR and C#, but not, I'm afraid, generalized constraints.

I'm now at Meta (formerly Facebook), and have implemented the generalized constraints feature (and more) for the programming language Hack that is used here.
